feat(kad): New provider record update strategy

Mivik opened the pull request:
## Description

In `MemoryStore`, the number of provider records per key is limited by `max_providers_per_key`. Former implementations keep provider records sorted by their distance to the key, and only keep those with the smallest distance. This strategy is vulnerable to Sybil attack, in which an attacker can flood the network with false identities in order to eclipse a key.

This commit change the strategy to simply keep old providers and ignore new ones. This new strategy however, can cause load imbalance, but can be mitigated by increasing `max_providers_per_key`.

In addition, old implementations failed to keep `provided` and `providers` in sync, and this commit fixes this issue.

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

Mivik committed:
feat(kad): Keep `provided` in sync with `providers` in `MemoryStore`

guillaumemichel commented on protocols/kad/src/record/store/memory.rs:
```diff
@@ -152,38 +152,45 @@ impl RecordStore for MemoryStore {
+        let position = providers
+            .iter()
+            .position(|p| key.distance(&provider) <= key.distance(&kbucket::Key::from(p.provider)))
+            .unwrap_or_else(|| providers.len());
```
Why do we want to store the provider records of peers whose identity is close to the key they provide in priority?

I see this was already the case in the past, but I think it is wrong. It make sense to have an upper bound in the number of provider records stored for a key, but this number doesn't have to be the magic `K`, it should be customizable. Storing in priority records from peers that are close to the key make censorship of keys easy, since an attacker could forge identities close to the key they want to eclipse, and advertise they have the key (and not serve content upon request). All previous providers will be dropped from the DHT, and the content cannot be discovered.

This can be tackled in a follow up PR.

Mivik replied:
> It make sense to have an upper bound in the number of provider records stored for a key, but this number doesn't have to be the magic K, it should be customizable.

It's already customizable (`max_providers_per_key`)

> Storing in priority records from peers that are close to the key make censorship of keys easy

Sure but please note that storing peers equally has the same problem (but less serious). As long as there's an upper bound in the number of provider records, the attacker can forge enough many identities to eclipse the key. I investigated `go-libp2p-kad-dht` and [it](https://github.com/libp2p/go-libp2p-kad-dht/blob/25d299e64f3cb9a3da337e9d4af1a1b722ae503d/providers/providers_manager.go#L271) seems that it does not impose explicit limit on the number of provider records.

Mivik replied:
Since this upper bound is not mentioned in the [specs](https://github.com/libp2p/specs/tree/master/kad-dht) either, I suggest having no limits in the number of providers per key by default. However if a limit is set and an excess provider is added to the store, we can randomly choose a provider record to replace. Does that look good to you?

guillaumemichel replied:
`rust-libp2p` follows the principle to [bound everything](https://github.com/libp2p/rust-libp2p/blob/master/docs/coding-guidelines.md#bound-everything), so let's keep a customizable upper bound 👍🏻 

IMO we should get rid of the XOR distance sorting because it isn't appropriate. I can see different solutions to decide which providers to keep:
1. Seniority: if we have enough providers, drop new providers. If an existing provider republishes a record for the same key, update the record. Good for censorship resistance, bad for load balancing.
2. Override: When a new provider is received, override a random existing provider with the new provider. Terrible to censorship resistance, good for load balancing.

I have a preference for the first approach since load balancing should be improved by increasing the `max_providers_per_key`, and there would be no way to compensate for censorship vulnerability. Moreover, the second approach would make the network more random and harder to debug.

Mivik replied:
Sounds good

Mivik committed:
refactor(kad): change provider record keep strategy of MemoryStore

Former implementations keep provider records sorted by their distance to
 the key, however this strategy is vulnerable to Sybil attack, in which
 an attacker can flood the network with false identities in order to
 eclipse a key.

This commit change the strategy to simply keep old providers and ignore
new ones. This new strategy however, can cause load imbalance, but can
be mitigated by increasing `max_providers_per_key`.

guillaumemichel committed:
minor corrections

guillaumemichel reviewed (approved):
Looks good to me! Thanks @Mivik for your contribution!

Can you also add an entry in the changelog and bump the version of `kad`?

Mivik committed:
chore(kad): Bump to 0.46.1

guillaumemichel committed:
Merge branch 'master' into master

guillaumemichel committed:
bumping libp2p version
